### PR TITLE
extract the func object from partial while scripting

### LINKED ISSUE #43520: [jit] cannot use functools.partial with TorchScript
## 🐛 Bug

TorchScript wants to derive a name from a function using `__name__`. `functools.partial` does not have `__name__` but has `name`. 

## To Reproduce

```py
from functools import partial

import torch


def sum_fun(x: torch.Tensor, dim: int = 0) -> torch.Tensor:
    return torch.sum(x, dim=dim)


sum_script = torch.jit.script(sum_fun) # works
sum_partial_script = torch.jit.script(partial(sum_fun, dim=-1)) # fails
```

## Expected behavior

Both work :)

## Environment

<details>
PyTorch version: 1.5.1
Is debug build: False
CUDA used to build PyTorch: 10.1

OS: CentOS Linux release 7.4.1708 (Core)  (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
Clang version: Could not collect
CMake version: version 2.8.12.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.5.1
[pip3] torchvision==0.6.0a0+35d732a
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] hamiltorch                0.3.1.dev3               pypi_0    pypi
[conda] mkl                       2020.2                      256    conda-forge
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.1.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38hcb8c335_0    conda-forge
[conda] numpy                     1.19.1           py38hbc911f0_0  
[conda] numpy-base                1.19.1           py38hfa32c7d_0  
[conda] pytorch                   1.5.1           py3.8_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.6.1                py38_cu101    pytorch
</details>

### PR DESCRIPTION
**Pull request by hlahkar**
Fixes #43520

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -15049,7 +15056,7 @@ def fn(string):
     def test_unicode_comments(self):
         @torch.jit.script
         def test(self, a):
-            # 🤷🤷🤷🤷
```
Comment (gmagogsfm) [inline test/test_jit.py:15052]: Unintended changes?

File: test/test_jit.py [hunk 2]
```diff
@@ -788,6 +788,13 @@ def foo(x):
         x = torch.rand(3, 4)
         self.assertEqual(scripted(x), foo(x))
 
+    def test_partial(self):
+        def sum_fun(x: torch.Tensor, dim: int = 0) -> torch.Tensor:
+            return torch.sum(x, dim=dim)
+        scripted_partial = torch.jit.script(functools.partial(sum_fun, dim=-1))
+        x = torch.ones(5)
+        self.assertEqual(scripted_partial(x), sum_fun(x))
```
Comment (gmagogsfm) [inline test/test_jit.py:796]: This seems a bit off to me. `partial` basically alters the default value of `dim`, but it is not reflected in scripted function.

File: torch/jit/_script.py [hunk 3]
```diff
@@ -898,6 +898,10 @@ def forward(self, input):
             obj, torch.jit._recursive.infer_methods_to_compile
         )
 
+    # for partial objects we need to extract the function for scripting
+    if isinstance(obj, functools.partial):
```
Comment (gmagogsfm) [inline torch/jit/_script.py:902]: Ditto:

IMHO, user's expectation would be scripting partial function, not the underlying original function. Let me know what you think.

**Review by gmagogsfm — CHANGES_REQUESTED** [review]
